### DISCUSSION TITLE
Minimizing "Reloading playlist" attempts / full retry?

### DISCUSSION THREAD

**Discussion by WolfganP** [streamlink/streamlink#4868] [Q&A] [answered]
Hi. It often happens to me that streamlink gets stuck on trying to reload the playlist. Ie:

```
streamlink\bin> .\streamlink.exe https://youtu.be/Bh1VCldzP5k
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.10.7
[cli][debug] Streamlink: 5.0.1
[cli][debug] Dependencies:
[cli][debug]  isodate: 0.6.0
[cli][debug]  lxml: 4.6.3
[cli][debug]  pycountry: 20.7.3
[cli][debug]  pycryptodome: 3.10.1
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.25.1
[cli][debug]  websocket-client: 1.1.0
[cli][debug] Arguments:
[cli][debug]  url=https://youtu.be/Bh1VCldzP5k
[cli][debug]  --loglevel=debug
[cli][debug]  --player="D:\Program Files\VideoLAN\VLC\vlc.exe"
[cli][debug]  --player-args=--qt-minimal-view
[cli][debug]  --default-stream=['720p', '720p_alt', '720p60', 'best']
[cli][debug]  --retry-max=2
[cli][debug]  --hls-playlist-reload-attempts=3
[cli][debug]  --hls-segment-threads=3
[cli][debug]  --ffmpeg-ffmpeg="streamlink\ffmpeg\ffmpeg.exe"
[cli][debug]  --generic-playlist-max=3
[cli][debug]  --generic-ignore-same-url=True
[cli][info] Found matching plugin youtube for URL https://youtu.be/Bh1VCldzP5k
[plugins.youtube][debug] Using video ID: Bh1VCldzP5k
[plugins.youtube][debug] This video is live.
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 144p (worst), 240p, 360p, 480p, 720p (best)
[cli][info] Opening stream: 720p (hls)
[cli][info] Starting player: "D:\Program Files\VideoLAN\VLC\vlc.exe"
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] First Sequence: 207; Last Sequence: 212
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 210; End Sequence: None
[stream.hls][debug] Adding segment 210 to queue
[stream.hls][debug] Adding segment 211 to queue
[stream.hls][debug] Adding segment 212 to queue
[stream.hls][debug] Segment 210 complete
[cli.output][debug] Opening subprocess: "X:\Program Files\VideoLAN\VLC\vlc.exe" --input-title-format https://youtu.be/Bh1VCldzP5k --qt-minimal-view -
[stream.hls][debug] Segment 211 complete
[stream.hls][debug] Segment 212 complete
[cli][debug] Writing stream to output
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
error: Error when reading from stream: Read timeout, exiting
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[stream.segmented][debug] Closing writer thread
[cli][info] Stream ended
[cli][info] Closing currently open stream...

```

I end the attempts by closing the player (vlc in this case). I tried to minimize the failure to stream with  `--retry-max=2` and `--hls-playlist-reload-attempts=3` but to not avail, and I wanted to make sure I'm not misinterpreting the cli options' actions (btw, it's not isolated to youtube, other streams fail the same way)

What also happens is that if I launch the same command line again, the stream restarts playing without errors. Is there any option I can use to command streamlink to do a full retry as if relaunched? (because it seems the playlist reloading is not the same full action)

Thanks a lot!

**Comment by bastimeyer** [comment, marked as answer]
The `EXT-X-ENDLIST` tag is missing from the (live) media playlist, hence the reload attempts.
https://github.com/streamlink/streamlink/issues/2198
https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.3.4

**Reply by bastimeyer** [reply]
`--hls-playlist-reload-attempts` only sets the number of reload attempts when reloading the playlist fails. This is not the case here however, and the reloads simply don't contain any new segments to queue. Then the regular `--stream-timeout` value of 60 seconds kicks in and the streams ends this way, ungracefully.

**Reply by WolfganP** [reply]
> `--hls-playlist-reload-attempts` only sets the number of reload attempts when reloading the playlist fails. This is not the case here however, and the reloads simply don't contain any new segments to queue. Then the regular `--stream-timeout` value of 60 seconds kicks in and the streams ends this way, ungracefully.

Thanks a lot for the explanation. I'll experiment with `--stream-timeout` then to try optimize these cases.
Regarding the full retry followup question, is there any cli option that may force refresh the stream the same way as if stopping / relaunching streamlink?

**Reply by bastimeyer** [reply]
> is there any cli option that may force refresh the stream the same way as if stopping / relaunching streamlink?

If the resolved and opened HLS playlist times out due to specific streaming provider logic, then it's up to the plugin to re-fetch an up-to-date playlist URL via a custom HLS subclass. Some plugins do this, but most don't.

If this hasn't been implemented for the plugin you're using, then all you can do is restart the whole Streamlink process. Once a stream has been resolved, there is no restart stream option. You can use a command line shell loop.
